src/common/mime.cc

```diff
@@ -15,6 +15,88 @@
 #include <pistache/http.h>
 #include <pistache/mime.h>
 
+/*
+ * This function parses a non-NULL terminated C string and interprets it as
+ * a float. The str must represent a number following the HTTP definition
+ * of Quality Values:
+ *
+ *     qvalue = ( "0" [ "." 0*3DIGIT ] )
+ *            / ( "1" [ "." 0*3("0") ] )
+ *
+ * https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1
+ */
+static bool str_to_qvalue(const char* str, float* qvalue, std::size_t* qvalue_len)
+{
+    constexpr char offset = '0';
+
+    *qvalue_len = 0;
+
+    // It is useless to read more than 6 chars, as the maximum allowed
+    // number of digits after the dot is 3, so n.nnn is 5.
+    // The 6th character is read to check if the user specified a qvalue
+    // with too many digits.
+    for (; *qvalue_len < 6; (*qvalue_len)++)
```

Comment:
Because you're already looping, probably you can parse the number as an integer in the loop as well. I understand that's also what @dennisjenkins75 meant: you simply do the classic `qvalue_int = qvalue_int * 10 + (str[i] - '0')`. Only one loop -> function simplified ; no precision loss involved by dealing with floating points.

Reply:
Maybe my brain simply doesn't want to collaborate, but could you please explain me how you would do this?